metrics: add new metrics to the new endpoint

NyaliaLui opened the pull request:
This patch adds several metrics to the new prometheus endpoint (exposed at "metrics2"). The following metrics were added:

* redpanda_kafka_consumer_group_committed_offset
  * Description: Group topic partition offset
  * Labels: group,topic,namespace,partition
  * Aggregation: shard
* redpanda_kafka_consumer_group_consumers
  * Description: Number of consumers in a consumer group
  * Labels: group
  * Aggregation: shard
* redpanda_kafka_consumer_group_topics
  * Description: Number of topics in a consumer group
  * Labels: group
  * Aggregation: shard
* redpanda_pandaproxy_request_errors_total
  * Description: Total number of request errors
  * Labels: none
  * Aggregation: shard
* redpanda_schema_registry_request_errors_total
  * Description: Total number of request errors
  * Labels: none
  * Aggregation: shard
* redpanda_shadow_indexing_errors_total
  * Description: Number of transmission & recieve errors
  * Labels: tx, rx
  * Aggregation: shard
* redpanda_rpc_request_errors_total
  * Description: Number of rpc errors
  * Labels: component
  * Aggregation: shard
* redpanda_rpc_latency_seconds
  * Description: Internal and Kafka RPC latency
  * Labels: component
  * Aggregation: shard

BenPope committed:
admin: Add second metrics endpoint

Expose another prometheus endpoint on route '/public_metrics'. This
endpoint should include a limited set of metrics, aggregated by default
that's useful for external and internal users who lack context into
the innerworkings of the project.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
metrics: Allow histogram_logform to be configurable

Signed-off-by: Ben Pope <ben@redpanda.com>

Commit:
cluster: add ntp metrics to new endpoint

This patch adds partition and topic level endpoint to the new prometheus
endpoint (exposed at "/public_metrics"). The following metrics were
added:

Partition Level Metrics:

* redpanda_kafka_max_offset:
    * Description: latest committed offset for the partition (i.e. the
    offset of the last message safely persisted on most replicas)
    * Labels: namespace, topic, partition, shard
    * Aggregation: shard
* redpanda_kafka_under_replicated_replicas:
    * Description: the number of under replicated replicas for the
      partition
    * Labels: namespace, topic, partition, shard
    * Aggregation: shard

Topic Level Metrics:
* redpanda_kafka_request_bytes_total:
    * Description: total number of bytes consumed/produced per topic
    * Labels: namespace, topic, partition,
    request ("produce" | "consume")
    * Aggregation: shard, partition
* redpanda_kafka_replicas:
    * Description: number of configured replicas per topic
    * Labels: namespace, topic, partition
    * Aggregation: shard, partition

Broker Level Metrics:
* redpanda_kafka_request_latency_seconds:
    * Description: latency of produce/consume requests per broker
    (measures from the moment a request is initiated on the partition to
    the moment when the response is fulfilled)
    * Labels: shard
    * Aggregation: shard

Commit:
cluster: add cluster level metrics via a new probe

This patch introduces a new probe that which is embedded in the
controller object. Only one instance is created per broker and it
matches the lifetime of the controller object. These metrics are
reported only from the controller leader (one node per cluster).

The following cluster level metrics were added:
* redpanda_cluster_brokers:
    * Description: number of configured brokers in the cluster
    * Label: shard
    * Aggreation: shard
* redpanda_cluster_partitions:
    * Description: number of partitions managed by the cluster (note
      that replicas are not counted as partitions and that the
      partitions of the controller topic are included)
    * Label: shard
    * Aggregation: shard
* redpanda_cluster_unavailable_partitions:
    * Description: number of unavailable partitions in the cluster (i.e.
      partitions that lack quorum among replicants)
    * Label: shard
    * Aggregation: shard

Commit:
pandaproxy: Add latency metric to public endpoint

This patch adds request latency metrics for pandaproxy and the schema
registry.

* repdanda_pandaproxy_request_latency_seconds
* redpanda_schema_registry_request_latency_seconds
    * Description: latency of requests processed by the respective
      server which includes the time spent waiting for resources to
      become available
    * Labels: shard, operation
    * Aggregation: shard, operation

This patch changes the server intialisation to thread through the name
of the server to the metric registration scope. Prior to this change
pandaproxy and schema_registry metrics where both reported under the
`redpanda_pandaproxy` prefix, while now they're reported separately.

NyaliaLui committed:
kafka/group: add new consumer group metrics

This patch adds consumer group metrics to the new prometheus endpoint
(exposed at "public_metrics"). The following metrics were added:

* redpanda_kafka_consumer_group_committed_offset
	* Description: Group topic partition offset
	* Labels: group,topic,partition,namespace
	* Aggregation: shard
* redpanda_kafka_consumer_group_consumers
	* Description: Number of consumers in a consumer group
	* Labels: group
	* Aggregation: shard
* redpanda_kafka_consumer_group_topics
	* Description: Number of topics in a consumer group
	* Labels: group
	* Aggregation: shard

BenPope commented on src/v/kafka/group_probe.h:
```diff
@@ -44,11 +47,66 @@ class group_offset_probe {
+            sm::make_gauge(
+            "committed_offset",
+            [this] { return _offset; },
+            sm::description("Group topic partition offset"),
```
```suggestion
            sm::description("Consumer group committed offset"),
```

NyaliaLui replied:
done

BenPope reviewed (commented):
Nice

Commit:
ssx: Add default seastar metric config utilities

This patch adds a new 'metrics.h' header that exposes some utilites
for interacting with the seastar metrics subsystem:
* public_metrics_handle: the seastar internal implementation handle
for metrics exposed on the '/public_metrics' endpoint
* report_default_histogram: converts an HDR histogram to a seastar
histogram with pre-configured defaults

NyaliaLui committed:
pandaproxy: add metrics for error count

This patch adds pandaproxy metrics to the new prometheus endpoint
(exposed at "public_metrics"). The following metrics were added:

* redpanda_pandaproxy_request_errors_total
	* Description: Total number of request errors
 	* Labels: none
        * Aggregation: shard
* redpanda_schema_registry_request_errors_total
        * Description: Total number of request errors
        * Labels: none
        * Aggregation: shard

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -55,4 +56,24 @@ probe::probe(
+        sm::description(ssx::sformat("Total number of {} request errors", group_name)),
+        {},
+        {sm::shard_label})},
+      sm::impl::default_handle() + 1);
```
I think this handle got a name in `ssx/metrics.h`

NyaliaLui replied:
I'll make this change when I rebase in the new changes from Vlad

NyaliaLui replied:
done

BenPope commented on src/v/pandaproxy/probe.h:
```diff
@@ -29,4 +29,16 @@ class probe {
 
+class error_probe {
+public:
+    error_probe(const ss::sstring& group_name);
```
```suggestion
    explicit error_probe(const ss::sstring& group_name);
```

NyaliaLui replied:
done

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,71 @@ errored_body(std::error_condition ec, ss::sstring msg) {
+inline std::unique_ptr<ss::httpd::reply>
+errored_body(std::error_condition ec, ss::sstring msg, error_probe& eprobe) {
+    auto rep = errored_body(ec, std::move(msg));
+    eprobe.increment_error_count();
```
Not all errors are equal. Should the status be checked for >=400, or >=500? Should we have a label for `3xx`, `4xx`, `5xx`? There are options, here.

NyaliaLui replied:
That's a good point. I think adding a label for different error groups makes sense here.

NyaliaLui replied:
I shall make the change.

NyaliaLui replied:
done

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,71 @@ errored_body(std::error_condition ec, ss::sstring msg) {
         return errored_body(
           reply_error_code::kafka_bad_request,
-          e.what()); // TODO BP: Yarr!!
+          e.what(), eprobe); // TODO BP: Yarr!!
```
I think my silly comment be dropped!

NyaliaLui replied:
done

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,71 @@ errored_body(std::error_condition ec, ss::sstring msg) {
+public:
     ss::sstring mime_type;
+
+    explicit exception_replier(ss::sstring mt, error_probe& eprobe)
```
```suggestion
    exception_replier(ss::sstring mt, error_probe& eprobe)
```

NyaliaLui replied:
done

NyaliaLui committed:
cloud_storage: add metrics for error count

This patch adds cloud_storage metrics to the new prometheus endpoint
(exposed at "public_metrics"). The following metrics were added:

* redpanda_shadow_indexing_errors_total
	* Description: Number of transmission & recieve errors
        * Labels: tx, rx
        * Aggregation: shard

NyaliaLui committed:
rpc: add error count to the new endpoint

This patch adds rpc metrics to the new prometheus endpoint
(exposed at "public_metrics"). The following metrics were added:

* redpanda_rpc_request_errors_total
       * Description: Number of rpc errors
       * Labels: component
       * Aggregation: shard

NyaliaLui committed:
rpc: add latency measurements to the new endpoint

This patch adds rpc latency metrics to the new prometheus endpoint
(exposed at "public_metrics"). The following metrics were added:

* redpanda_rpc_latency_seconds
	* Description: Internal and Kafka RPC latency
	* Labels: component
	* Aggregation: shard

BenPope commented on src/v/kafka/group_probe.h:
```diff
@@ -44,11 +52,86 @@ class group_offset_probe {
+            [this, &offsets] {
+              // Capture list of unique topic partitions
+              // by storing them in a set
+              for (auto &elem: offsets) {
+                _topics.insert(elem.first);
+              }
+
+              return _topics.size();
```
I think this is what we want:
```suggestion
              return _offsets.size();
```
Since the key is the same for the underlying map and the set.

But holding topics over time means that we're counting every topic that was ever registered, rather than the current view (accounting for removed topics). Is that what you intended?

NyaliaLui replied:
> Since the key is the same for the underlying map and the set.

Ah right, there are not duplicate keys in the offset map. I must've been tired when writing that in. I'll make the change.

> But holding topics over time means that we're counting every topic that was ever registered, rather than the current view (accounting for removed topics). Is that what you intended?

No that is not my intention, I also thought about this overnight and I'll remove the topics container in the next push.

NyaliaLui replied:
Done. Changes were pushed to a separate PR at https://github.com/redpanda-data/redpanda/pull/5349

BenPope commented on src/v/kafka/group_probe.h:
```diff
@@ -44,11 +52,86 @@ class group_offset_probe {
+private:
+    member_map& _members;
+    static_member_map& _static_members;
+    absl::node_hash_set<KeyType> _topics;
```
I don't think this member is needed based on the previous comment.

`_members` and `_static_members` held as a member reference, whereas `_offsets` is a capture from `setup_metrics`. I think holding `_offsets` here by reference at the point of construction more clearly represents the lifetimes..

NyaliaLui replied:
I agree, I'll make the change.
I was going to experiment with passing _members and _static_members through setup_metrics but I think using a member reference makes the code more readable.

NyaliaLui replied:
Done. Changes were pushed to a separate PR at https://github.com/redpanda-data/redpanda/pull/5349

BenPope commented on src/v/kafka/group_probe.h:
```diff
@@ -12,18 +12,26 @@
 
     void setup_metrics(
-      const kafka::group_id& group_id, const model::topic_partition& tp) {
+      const kafka::group_id& group_id, const model::topic_partition& tp, model::topic_namespace& tp_ns) {
```
I'd expect to see an `ntp` passed instead of a `tp` and `tp_ns`.

BenPope replied:
I think I followed it through correctly, but isn't this the `tp_ns` of the `model::kafka_group_nt`?

NyaliaLui replied:
> I think I followed it through correctly, but isn't this the `tp_ns` of the `model::kafka_group_nt`?

I'll check, but I also think we can remove namespace from this metric since using the consumer group API guarantees the namespace will always be kafka. I'll make a comment in the RFC document to collect opinions.

BenPope replied:
Yeah, let's just remove namespace.

NyaliaLui replied:
Done. Changes were pushed to a separate PR at https://github.com/redpanda-data/redpanda/pull/5349

BenPope commented on src/v/kafka/server/group.h:
```diff
@@ -534,7 +538,7 @@ class group final : public ss::enable_lw_shared_from_this<group> {
               std::move(tp),
               std::make_unique<offset_metadata_with_probe>(
-                std::move(md), _id, tp, _enable_group_metrics));
+                std::move(md), _id, tp, _tp_ns, _enable_group_metrics));
```
Not your code, but the `tp` passed here may have been moved.

NyaliaLui replied:
Looks like tp is still there :)

Changes were pushed to a separate PR at https://github.com/redpanda-data/redpanda/pull/5349

BenPope commented on src/v/kafka/server/group.h:
```diff
@@ -549,7 +553,7 @@ class group final : public ss::enable_lw_shared_from_this<group> {
               std::move(tp),
               std::make_unique<offset_metadata_with_probe>(
-                std::move(md), _id, tp, _enable_group_metrics));
+                std::move(md), _id, tp, _tp_ns, _enable_group_metrics));
```
Not your code, but the `tp` passed here may have been moved.

NyaliaLui replied:
Looks like tp is still there :)

Changes were pushed to a separate PR at https://github.com/redpanda-data/redpanda/pull/5349

BenPope commented on src/v/kafka/server/tests/group_test.cc:
```diff
@@ -51,6 +51,7 @@ static group get() {
       conf,
       nullptr,
       make_backward_compatible_serializer(),
+      model::topic_namespace{model::ns{"kafka"}, model::topic{"sample"}},
```
```suggestion
      model::topic_namespace{model::kafka_namespace, model::topic{"sample"}},

NyaliaLui replied:
This change was removed since I also removed the namespace label from committed_offset metric.

Changes were pushed to a separate PR at https://github.com/redpanda-data/redpanda/pull/5349

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -53,4 +53,50 @@ probe::probe(
+        {status_label("3xx")})
+        .aggregate({sm::shard_label}),
+
+        sm::make_counter(
+        "request_errors_total",
+        [this] { return _5xx_count + _4xx_count + _3xx_count; },
+        sm::description(ssx::sformat("Total number of {} errors", group_name)),
+        {})
+        .aggregate({sm::shard_label})
```
This is redundant, the PromQL is just `sum without(status)(redpanda_<group>_request_errors_total)`

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5422

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -53,4 +53,50 @@ probe::probe(
+  , _4xx_count(0)
+  , _3xx_count(0)
+  , _public_metrics(ssx::public_metrics_handle) {
+    if (config::shard_local_cfg().disable_metrics()) {
```
This should be `disable_public_metrics()`

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5422

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,79 @@ errored_body(std::error_condition ec, ss::sstring msg) {
+
+    // See seastar/http/reply.hh::reply::status_type for a list
+    // of supported error status codes.
+    if (rep->_status >= status_type{500}) eprobe.increment_5xx();
+    else if (rep->_status >= status_type{400}) eprobe.increment_4xx();
+    else if (rep->_status >= status_type{300}) eprobe.increment_3xx();
```
I think the braces should be added for consistency (and this will be reformatted by clang-format)

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5422

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,79 @@ errored_body(std::error_condition ec, ss::sstring msg) {
+    return rep;
+}
+
 inline std::unique_ptr<ss::httpd::reply>
 errored_body(std::error_code ec, ss::sstring msg) {
     return errored_body(make_error_condition(ec), std::move(msg));
 }
```
This is now unused

NyaliaLui replied:
I'm using a different approach now.
Changes at  https://github.com/redpanda-data/redpanda/pull/5422

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,79 @@ errored_body(std::error_condition ec, ss::sstring msg) {
+    return errored_body(make_error_condition(ec), std::move(msg), eprobe);
+}
+
 inline std::unique_ptr<ss::httpd::reply> unprocessable_entity(ss::sstring msg) {
     return errored_body(
       make_error_condition(reply_error_code::kafka_bad_request),
       std::move(msg));
 }
```
This is unused

NyaliaLui replied:
I'm using a different approach.
Changes at https://github.com/redpanda-data/redpanda/pull/5422

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -76,50 +77,79 @@ errored_body(std::error_condition ec, ss::sstring msg) {
 
+inline std::unique_ptr<ss::httpd::reply>
+errored_body(std::error_condition ec, ss::sstring msg, error_probe& eprobe) {
+    auto rep = errored_body(ec, std::move(msg));
```
I think the overload of `errored_body` that's called here, should be inlined, or there's a good chance it can be called and the probe missed off.

NyaliaLui replied:
I'm using a different approach than what is here.
Changes at https://github.com/redpanda-data/redpanda/pull/5422

BenPope commented on src/v/cloud_storage/probe.cc:
```diff
@@ -12,12 +12,15 @@
-remote_probe::remote_probe(remote_metrics_disabled disabled) {
+remote_probe::remote_probe(remote_metrics_disabled disabled)
+  : _public_metrics(ssx::public_metrics_handle) {
     if (disabled) {
```
There's another flag for `disable_public_metrics`, so I think the early return is wrong.

NyaliaLui replied:
Note to self:

disable_public_metrics comes from `src/v/config/configuration` where as the cloud storage subsystem uses the configuration from `src/v/net/transport`.

I'll have to scan through the code base more to see how to reconcile this difference.

NyaliaLui replied:
Done. I had to add public metrics to the configuration container in src/v/net

Changes at https://github.com/redpanda-data/redpanda/pull/5367

BenPope commented on src/v/cloud_storage/probe.cc:
```diff
@@ -95,6 +98,25 @@ remote_probe::remote_probe(remote_metrics_disabled disabled) {
+        sm::make_counter(
+          "errors_total",
+          [this] { return get_failed_transmissions(); },
+          sm::description("Number of transmission & recieve errors"),
```
```suggestion
          sm::description("Number of transmit errors"),

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5367

BenPope commented on src/v/cloud_storage/probe.cc:
```diff
@@ -95,6 +98,25 @@ remote_probe::remote_probe(remote_metrics_disabled disabled) {
+        sm::make_counter(
+          "errors_total",
+          [this] { return get_failed_recieves(); },
+          sm::description("Number of transmission & recieve errors"),
```
```suggestion
          sm::description("Number of receive errors"),

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5367

BenPope commented on src/v/cloud_storage/probe.cc:
```diff
@@ -95,6 +98,25 @@ remote_probe::remote_probe(remote_metrics_disabled disabled) {
+          .aggregate({sm::shard_label}),
+        sm::make_counter(
+          "errors_total",
+          [this] { return get_failed_recieves(); },
```
```suggestion
          [this] { return get_failed_downloads() + get_failed_manifest_downloads(); },

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5367

BenPope commented on src/v/cloud_storage/probe.cc:
```diff
@@ -95,6 +98,25 @@ remote_probe::remote_probe(remote_metrics_disabled disabled) {
+      {
+        sm::make_counter(
+          "errors_total",
+          [this] { return get_failed_transmissions(); },
```
```suggestion
          [this] { return get_failed_uploads() + get_failed_manifest_uploads(); },

NyaliaLui replied:
Done. 
Changes at https://github.com/redpanda-data/redpanda/pull/5367

BenPope commented on src/v/cloud_storage/probe.h:
```diff
@@ -164,8 +176,13 @@ class remote_probe {
     uint64_t _cnt_bytes_sent{0};
     /// Number of bytes being successfully received from S3
     uint64_t _cnt_bytes_received{0};
+    /// Number of failed requests on the transmission path
+    uint64_t _cnt_failed_tx{0};
+    /// Number of failed requests on the recieve path
+    uint64_t _cnt_failed_rx{0};
```
I think we can aggregate the existing metrics

NyaliaLui replied:
Good point.

NyaliaLui replied:
Done.
Changes at https://github.com/redpanda-data/redpanda/pull/5367

BenPope commented on src/v/net/probes.cc:
```diff
@@ -92,6 +96,31 @@ void server_probe::setup_metrics(
+
+    ss::sstring proto_name(proto);
+
+    // Only register the metric for one proto
+    // otherwise we get double registration problems
```
This class used to be agnostic of kafka and internal, those are injected. The protos are `internal_rpc` and `kafka_rpc`, so the names aren't quite what we want.

I think we should reuse  `_service_errors`, which is the same metric, but agnostic of the proto.

Maybe it's best to remove the conditional and the new probes and do something like:

```c++
std::string_view proto_name(proto);
proto_name.remove_suffix(4);
...
component_label(proto_name);
```

NyaliaLui replied:
I don't entirely understand your comment but I agree that we can find a different approach. I'll look into this more.

BenPope replied:
I believe there's an instance of the class for internal and an instance for kafka. Now both instances contain metrics for the instance (service_errors), for internal, and for kafka, which is why there was double registration.

BenPope commented on src/v/redpanda/application.cc:
```diff
@@ -341,6 +343,26 @@ void application::setup_metrics() {
+            "latency_seconds",
+            sm::description("Internal RPC latency"),
+            {component_label("internal")},
+            [this] { return _rpc.local().rpc_hist()->seastar_histogram_logform(); })
+            .aggregate({sm::shard_label}),
```
Can'r we aggregate the existing rpc latency metrics and use the ssx::metrics::default_histgram?

NyaliaLui replied:
With that approach wouldn't the metrics between internal_rpc and kafka_rpc be duplicates of each other? I'll experiment to confirm.

BenPope commented on src/v/net/server.h:
```diff
@@ -178,6 +179,8 @@ class server {
 
     const server_configuration cfg; // NOLINT
     const hdr_hist& histogram() const { return _hist; }
+    ss::lw_shared_ptr<hdr_hist> rpc_hist() { return _rpc_hist; }
+    hdr_hist& kafka_hist() { return _kafka_hist; }
```
This doesn't feel right, can't we reuse `_hist`?

NyaliaLui replied:
With that approach wouldn't the metrics between internal_rpc and kafka_rpc be duplicates of each other? I'll experiment to confirm.

BenPope replied:
I think there's an instance of this class for kafka, and one for internal.

BenPope commented:
Can you make sure you run clang-format.

BenPope commented:
Can you ensure that the metrics are disabled with `disable_public_metrics()` and not with `disable_metrics()`

mmedenjak commented:
I guess we can close this now in favour of https://github.com/redpanda-data/redpanda/pull/5349?

EDIT: I guess the PR for dev is just a subset of this one, right?

NyaliaLui commented:
> I guess we can close this now in favour of [redpanda-data#5349](https://github.com/redpanda-data/redpanda/pull/5349)?
> 
> EDIT: I guess the PR for dev is just a subset of this one, right?

Yes, the PR for dev is a subset. I can go ahead and close it now though. I'm simply using this PR as a checklist at this point.
